### PR TITLE
Allow Prow GKE Service Accounts to assume AWS roles

### PR DESCRIPTION
**Pull request by upodroid**
/cc @xmudrii @ameukam 

This is the AWS piece of getting Prow to run pods on EKS. The GKE bit is to adjust the manifests to have the correct env and a projected service account token with the correct claims.

/sig testing
/sig k8s-infra

### CODE REVIEW

**Comment by xmudrii** [comment]
/hold
some additional changes might be needed

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+resource "aws_iam_openid_connect_provider" "k8s_prow" {
+  url             = "https://container.googleapis.com/v1/projects/k8s-prow/locations/us-central1-f/clusters/prow"
+  client_id_list  = ["sts.amazonaws.com"]
+  thumbprint_list = ["08745487e891c19e3078c1f2a07e452950ef36f6"]
+}
+
+# s3writer iam role for artifacts management
+# We allow the kubernetes service account to assume this role
+resource "aws_iam_role" "eks_admin" {
+  name = "EKS-Admin"
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/prow.tf:30]: Should we call this `Prow-EKS-Admin`?

Comment (upodroid) [reply infra/aws/terraform/prow-build-cluster/prow.tf]: Fixed

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        "Effect" : "Allow",
+        "Principal" : {
+          "Federated" : aws_iam_openid_connect_provider.k8s_prow
+        },
+        "Action" : "sts:AssumeRoleWithWebIdentity",
+        "Condition" : {
+          "StringEquals" : {
+            "container.googleapis.com/v1/projects/k8s-prow/locations/us-central1-f/clusters/prow:sub" : [
+                // Prow Services that interact with Kubernetes can be found at https://github.com/kubernetes/test-infra/blob/master/.ko.yaml
+                "system:serviceaccount:default:deck",
```
Comment (upodroid) [inline infra/aws/terraform/prow-build-cluster/prow.tf:44]: @kubernetes/test-infra-maintainers Are these the right service accounts for pods that interact with kubernetes clusters?

File: infra/aws/terraform/prow-build-cluster/eks.tf [hunk 3]
```diff
@@ -30,6 +30,17 @@ module "eks" {
   # Manage aws-auth ConfigMap.
   manage_aws_auth_configmap = true
 
+  # Configure aws-auth
+  aws_auth_roles = [
+    {
+      "rolearn"  = aws_iam_role.eks_admin.id
+      "username" = "eks-admin"
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/eks.tf:37]: Where this value comes from? Should that eventually be name of IAM role or that's something completely else? I'm still trying to understand how this ConfigMap works at all.

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+*/
+
+// This IAM configuration allows Prow GKE Clusters to assume a role on AWS
+
+
+# Recognize federated identities from the prow trusted cluster
+resource "aws_iam_openid_connect_provider" "k8s_prow" {
+  url             = "https://container.googleapis.com/v1/projects/k8s-prow/locations/us-central1-f/clusters/prow"
+  client_id_list  = ["sts.amazonaws.com"]
+  thumbprint_list = ["08745487e891c19e3078c1f2a07e452950ef36f6"]
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/prow.tf:24]: Maybe leave a comment where this value comes from (if not sensitive).

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+# s3writer iam role for artifacts management
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/prow.tf:27]: This comments needs an update. :)

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This IAM configuration allows Prow GKE Clusters to assume a role on AWS
+
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/prow.tf:18]: Very tiny nitpick:
```suggestion
# This IAM configuration allows Prow GKE Clusters to assume a role on AWS
```

File: infra/aws/terraform/prow-build-cluster/eks.tf (hunk 3 above)
Comment (upodroid) [reply infra/aws/terraform/prow-build-cluster/eks.tf:37]: This is an arbitary string. https://github.com/kubernetes-sigs/aws-iam-authenticator#crd-alpha

File: infra/aws/terraform/prow-build-cluster/prow.tf (hunk 4 above)
Comment (upodroid) [reply infra/aws/terraform/prow-build-cluster/prow.tf:24]: No, it is the thumprint for *.googleapis.com

https://awscli.amazonaws.com/v2/documentation/api/latest/reference/iam/update-open-id-connect-provider-thumbprint.html
https://cloud.google.com/anthos/clusters/docs/multi-cloud/aws/how-to/use-workload-identity-aws

File: infra/aws/terraform/prow-build-cluster/prow.tf (hunk 6 above)
Comment (upodroid) [reply infra/aws/terraform/prow-build-cluster/prow.tf:18]: Fixed

File: infra/aws/terraform/prow-build-cluster/prow.tf (hunk 4 above)
Comment (TerryHowe) [reply infra/aws/terraform/prow-build-cluster/prow.tf:24]: Ah, trusted cert, but this is OIDC

File: infra/aws/terraform/prow-build-cluster/prow.tf [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        "Condition" : {
+          "StringEquals" : {
+            "container.googleapis.com/v1/projects/k8s-prow/locations/us-central1-f/clusters/prow:sub" : [
+                // https://github.com/kubernetes/test-infra/tree/master/config/prow/cluster 
+                // all services that load kubeconfig should be listed here
+                "system:serviceaccount:default:deck",
+                "system:serviceaccount:default:config-bootstrapper",
+                "system:serviceaccount:default:crier",
+                "system:serviceaccount:default:sinker",
+                "system:serviceaccount:default:prow-controller-manager"
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/prow.tf:48]: This is missing comma.
```suggestion
                "system:serviceaccount:default:prow-controller-manager",
```

Comment (upodroid) [reply infra/aws/terraform/prow-build-cluster/prow.tf]: Fixed 

/label tide/merge-method-squash

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

The change is reconciled.

**Pull request merged**